### PR TITLE
Restructure code

### PR DESCRIPTION
**Pull request by davelosert**
In this pr, we want to tackel several tasks:

- [x] Switch to typescript
- [x] Split all rules across several files instead of having it all in `node.js`
- [x] Let `browser.js` and `node.js` extend a default-configuration rather than having `browser.js` extend `node.js`
- [x] Use a new format (BetterRules) for rules that allow a more precise syntax 
- [x] ~Allow functions as configuration arguments to be more dynamic with conditional configs (e.g. typescript vs. javascript)~
- [x] Writes some initial hooks (`usePlugin`, `useConditionalTs`) to use
- [x] ~Write a hook to use `unicorn`-overwrites in core rules~

---

Suggested squash commit message:

---

- Restructure rules into separate files
- Migrate package to typescript
- Create better rule configuration interface and utilities
- Export customization functionality and rules from `main`

BREAKING CHANGE:

- `main` does not export the config anymore. You have to explicitly extend `es/node` or `es/browser`
- Add some new rules

### CODE REVIEW

**Comment by davelosert** [comment]
@goloroden , @yeldiRium : this is ready for review.
The build failure is due to a failing dependency check with the package `cyclist` which is a sub-sub-sub-dependency of `semantic-release-configuration`. This should have actually failed already there - so I don't quite get why is only failing here.

However, as this will probably be fixed elsewhere, you could already start reviewing my changes here.

**Comment by yeldiRium** [comment]
> This should have actually failed already there - so I don't quite get why is only failing here.

As far as I can see in the failing action the problem is a licence incompatibility with us, which is due to cyclist v0.2.2 not having declared any license. We can safely add it as having an MIT license though, since they added the MIT license in a later commit. I'll add it to the license configuration

File: lib/isInstalled.ts [hunk 1]
```diff
@@ -1,26 +1,26 @@
...
     try {
-      // eslint-disable-next-line global-require
-      require(name);
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      import(name);
 
       // Theoretically, ex is not needed here, but Atom uses an older Node.js
       // version internally which is not able to deal with optional catch
       // bindings. Hence we disable this rule here instead of removing the
       // non-needed variable.
```
Comment (yeldiRium) [inline lib/isInstalled.ts:14]: I assume we can remove this and the caught exception below xD I would be very surprised if this is still relevant.

File: lib/pickRuleFor.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import { Language } from './Language';
```
Comment (yeldiRium) [inline lib/pickRuleFor.ts:1]: Let's rename this file to `pickRulesFor.ts`, so that the file name is identical to the exported function.

File: lib/Language.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+type Language = 'typeScript' | 'javaScript';
```
Comment (yeldiRium) [inline lib/Language.ts:1]: I believe we usually treat `typescript` and `javascript` as one word each, so i would not capitalize the `s` in the middle for consistency.

File: lib/betterRules/compile.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import { kebabCase } from 'lodash';
+import { Linter } from 'eslint';
+import { BetterRulesEntry, BetterRulesRecord } from './BetterRulesRecord';
+
+const compileRuleName = (ruleName: string): string => {
+  if (ruleName.includes('/')) {
```
Comment (yeldiRium) [inline lib/betterRules/compile.ts:6]: What do you think about moving the `fixReactEs6Rule` functionality here? It would make the `defaultConfiguration` a bit cleaner and hide the weirdness of formatting the rule names. Or could that conflict with the names of other rules?

File: lib/defaultConfiguration.ts [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  ...extended,
+  ...mochaRules,
+  ...unicorn,
+  ...createSharedRulesFor({ language: 'javaScript' })
+});
+
+// The ruleName of "react/prefer-es6-class" is wrongly converted to kebab-case
+// as lodash's kebab function treats the "6" as its own word, resulting in prefer-es-6-class.
+// The workaround below fixes this.
+const fixReactEs6Rule = (reactRulesRecord: Linter.RulesRecord): Linter.RulesRecord =>
```
Comment (yeldiRium) [inline lib/defaultConfiguration.ts:57]: Same as above, maybe consider moving this to `compile`?

File: package.json [hunk 6]
```diff
@@ -26,26 +26,37 @@
...
+    },
+    {
+      "name": "David Losert",
+      "email": "david.losert@gmx.de"
     }
   ],
   "private": false,
-  "main": "node.js",
-  "types": "",
+  "main": "build/lib/defaultConfig.js",
```
Comment (yeldiRium) [inline package.json:36]: Not really sure what I prefer here, but this could stay as `node.js`, couldn't it? `node.js` exports the same things `defaultConfig.js` does and it is the more obvious entrypoint to the library.

Also I don't think we need to export types from this package, since it is not really intended for programmatic consumption. And even if it were, we'd have to export `build/lib/index.d.ts`, not the types from `defaultConfig`.

File: test/unit/betterRules/compileTests.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import { assert } from 'assertthat';
+import { BetterRulesRecord } from '../../../lib/betterRules/BetterRulesRecord';
+import { compile } from '../../../lib/betterRules/compile';
+
+suite('rulesCompiler', (): void => {
+  test('compiles a rule with a  boolean to a deactivated rule.', async (): Promise<void> => {
```
Comment (yeldiRium) [inline test/unit/betterRules/compileTests.ts:6]: There a redundant space in here.

```suggestion
  test('compiles a rule with a boolean to a deactivated rule.', async (): Promise<void> => {
```

File: test/unit/betterRules/compileTests.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    const rule: BetterRulesRecord = {
+      noBraces: false
+    };
+
+    const result = compile(rule);
+
+    assert.that(result['no-braces']).is.not.undefined();
+  });
+
+  test('leaves plugin-names in tact when converting.', async (): Promise<void> => {
```
Comment (yeldiRium) [inline test/unit/betterRules/compileTests.ts:26]:
```suggestion
  test('leaves plugin-names intact when converting.', async (): Promise<void> => {
```

**Review by yeldiRium — CHANGES_REQUESTED** [review]
Amazing work! I am really impressed with how tidy this has turned out. I believe this will make the maintenance of the rule-set much easier in the future.

I have some questions, nitpicks and improvements, though :D

File: package.json (hunk 6 above)
Comment (davelosert) [reply package.json:36]: You are right.
But I think I would prefer exporting `build/lib/index.ts` as entry point.
This allows for programatically extending the rules and also using the `betterRules` package elsewhere. WDYT?

(Even though it's not intended yet, having the possibility seems nice)

File: lib/betterRules/compile.ts (hunk 4 above)
Comment (davelosert) [reply lib/betterRules/compile.ts:6]: It was there initially.

However - I kind of want to treat the `betterRules` Folder as if it was a generic plugin.
And this is a special 'bug' for a special package (`react`) that we are using.

So I found the right place was actually 'outside'.

I was thinking about providing an option to compile to be able to `postProcess` rules and then put it there. But since, at least right now, it is only this one single rule, that would be a bit premature imho.

Long story short:
For now, I'd preffer to leave it outside of the `betterRules` package. Understandable?

File: lib/defaultConfiguration.ts (hunk 5 above)
Comment (davelosert) [reply lib/defaultConfiguration.ts:57]: See my comment above.

File: lib/Language.ts (hunk 3 above)
Comment (davelosert) [reply lib/Language.ts:1]: As written in Slack, this is the technically correct form (as it is `JavaScript` and `TypeScript`), so we leave it here 🙂

File: lib/pickRuleFor.ts (hunk 2 above)
Comment (davelosert) [reply lib/pickRuleFor.ts]: Done - good catch. 🙂

File: lib/isInstalled.ts (hunk 1 above)
Comment (davelosert) [reply lib/isInstalled.ts]: It's gone. 👍🏻

File: lib/betterRules/compile.ts (hunk 4 above)
Comment (yeldiRium) [reply lib/betterRules/compile.ts:6]: Understandable :+1:

File: package.json (hunk 6 above)
Comment (yeldiRium) [reply package.json:36]: I like that solution too. So people using this as an eslint config package extend `es/node` as they did before, but if they want to use our functionality, they can import the package as `import {} from 'eslint-config-es`. Good solution :+1:

**Review by yeldiRium — APPROVED** [review]

**Review by yeldiRium — APPROVED** [review]

**Review by yeldiRium — DISMISSED** [review]

File: lib/Language.ts (hunk 3 above)
Comment (goloroden) [reply lib/Language.ts:1]: I double-checked this, and @yeldiRium is right – everywhere else where we had to make a difference between these two languages, we used the all-lowercase version, as if it was a single word. It is correct that the official spelling is with the uppercase `S`, but for the sake of consistency we should turn it into all-lowercase as well.

@davelosert Can you please adjust this once again? Sorry for the unnecessary work, this was clearly my fault.

File: lib/betterRules/BetterRulesRecord.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+type BetterRulesEntry = false | any[];
+type BetterRulesKey = string;
+
+// This is named "BetterRulesRecord" because we replace the ESLint Type Linter.RulesRecord
```
Comment (goloroden) [inline lib/betterRules/BetterRulesRecord.ts:4]:
```suggestion
// This is named "BetterRulesRecord" because we replace the ESLint built-in type "Linter.RulesRecord".
```

File: lib/defaultConfiguration.ts [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  extended,
+  mochaRules,
+  react,
+  typeScript,
+  unicorn
+} from './rules';
+
+const parserOptions: Linter.ParserOptions = {
+  sourceType: 'script',
+  ecmaVersion: 12,
```
Comment (goloroden) [inline lib/defaultConfiguration.ts:19]: Is there an option to have this as "stable" or "latest" or something like that, so that we don't have to adjust this from year to year?

File: lib/defaultConfiguration.ts [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+const globals = {};
+
+const plugins = [ 'eslint-comments', 'extended', 'mocha', 'unicorn' ];
+const settings = {};
+
+if (isInstalled('react')) {
+  plugins.push('react');
+
+  // @ts-expect-error react is actuall set by the plugin
```
Comment (goloroden) [inline lib/defaultConfiguration.ts:41]:
```suggestion
  // @ts-expect-error react is actually set by the plugin.
```

File: lib/rules/react.ts [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    ignoreCase: true,
+    requiredFirst: true,
+    sortShapeProp: true,
+    noSortAlphabetically: false
+  }],
+  stateInConstructor: [ 'always' ],
+  staticPropertyPlacement: false,
+  stylePropObject: [],
+  voidDomElementsNoChildren: [],
+
```
Comment (goloroden) [inline lib/rules/react.ts:68]: Why this empty line, and why having two distinct blocks for JSX- and non-JSX-rules?

File: lib/rules/sharedCoreTypeScript.ts [hunk 13]
```diff
@@ -0,0 +1,200 @@
+import { BetterRulesRecord } from '../betterRules';
+import { Language } from '../Language';
+import { pickRulesFor } from '../pickRulesFor';
+
+const createSharedRulesFor = ({ language }: { language: Language }): BetterRulesRecord => {
+  // CamelCase is a corner case - as for typescript, we only want it for javascript,
+  // but for typescript we want it deactivated and replaced by namingConvention
```
Comment (goloroden) [inline lib/rules/sharedCoreTypeScript.ts:7]:
```suggestion
  // camelCase is a corner case - as for TypeScript, we only want it for JavaScript,
  // but for TypeScript we want it deactivated and replaced by namingConvention
```

File: lib/rules/sharedCoreTypeScript.ts [hunk 14]
```diff
@@ -0,0 +1,200 @@
+import { BetterRulesRecord } from '../betterRules';
```
Comment (goloroden) [inline lib/rules/sharedCoreTypeScript.ts:1]: This file should then be named `sharedCoreTypescript.ts`.

File: lib/rules/unicorn.ts [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  noDocumentCookie: [],
+  noForLoop: [],
+  noHexEscape: [],
+  noInstanceofArray: [],
+  noInvalidRemoveEventListener: [],
+  noLonelyIf: [],
+  noNestedTernary: [],
+  noNewArray: [],
+
+  // The noNewBuffer rule does the same thing as ESLint's builtIn
```
Comment (goloroden) [inline lib/rules/unicorn.ts:60]:
```suggestion
  // The noNewBuffer rule does the same thing as ESLint's built-in
```

File: lib/rules/unicorn.ts [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+  preferStringReplaceAll: [],
+  preferStringSlice: [],
+  preferStringStartsEndsWith: [],
+  preferStringTrimStartEnd: [],
+
+  // This rule is turned off since it not only applies to assignments, but to
+  // expressions using await/throw/yield as well. We have opened an issue to try
+  // to remedy this: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1079
+  preferTernary: false,
+
```
Comment (goloroden) [inline lib/rules/unicorn.ts:119]: Why this empty line?

File: node.js [hunk 17]
```diff
@@ -1,1081 +1,5 @@
 'use strict';
 
-const path = require('path');
+const defaultEslintConfiguration = require('./build/lib/defaultConfiguration');
```
Comment (goloroden) [inline node.js:3]: Shouldn't the variable just be named `defaultConfiguration`, as in the `browser.js` file as well?

File: test/integration/dummyTs/dummy.ts [hunk 18]
```diff
@@ -0,0 +1,5 @@
+// We have to keep this file around to make the @typescript-eslint parser happy for tests
+// It is not actually used as we are using strings for our tests - but it avoids the error
```
Comment (goloroden) [inline test/integration/dummyTs/dummy.ts:2]:
```suggestion
// It is not actually used as we are using strings for our tests - but it avoids the error.
```

File: test/integration/dummyTs/dummyTsx.tsx [hunk 19]
```diff
@@ -0,0 +1,5 @@
+// We have to keep this file around to make the @typescript-eslint parser happy for tests
+// It is not actually used as we are using strings for our tests - but it avoids the error
```
Comment (goloroden) [inline test/integration/dummyTs/dummyTsx.tsx:2]:
```suggestion
// It is not actually used as we are using strings for our tests - but it avoids the error.
```

File: test/integration/esLintTester.ts [hunk 20]
```diff
@@ -0,0 +1,35 @@
+import { ESLint } from 'eslint';
+import path from 'path';
+import * as defaultConfig from '../../lib/defaultConfiguration';
```
Comment (goloroden) [inline test/integration/esLintTester.ts:3]:
```suggestion
import * as defaultConfiguration from '../../lib/defaultConfiguration';
```

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @davelosert 👋 

I also took a look at your PR, and I can only confirm what @yeldiRium had already said: Especially the definitions of the rules become so much more readable 🎉

Great work!
(Oh, and also a great idea to export the `betterRule` part.)

What I have not yet reviewed is the `README.md` file, and the workflows for GitHub Actions. However, both shoudn't be a blocker for meging.

File: lib/rules/sharedCoreTypeScript.ts (hunk 13 above)
Comment (davelosert) [reply lib/rules/sharedCoreTypeScript.ts]: Changing the first word to `camelCase` won't work as comments have to start with an upper case letter (that's the rules ;)). So I accpeted this suggestion but will have to change that again with my next commit.

File: lib/Language.ts (hunk 3 above)
Comment (davelosert) [reply lib/Language.ts]: No problem. I switched it back to lowercase.

File: lib/defaultConfiguration.ts (hunk 10 above)
Comment (davelosert) [reply lib/defaultConfiguration.ts]: There is `latest` which I used now.
However, due to an error in the types-package, typescript complains about it.

I fixed it with a `@ts-expect-error` comment now. Might send a PR to the `types` Repo later this week so we can remove the comment.

File: lib/rules/react.ts (hunk 12 above)
Comment (davelosert) [reply lib/rules/react.ts]: No reason.

I guess the jsx-separation was the sort-order before and I did not realize it.

I now put all jsx-rules into the upper block where they belong alphabetically.

File: lib/rules/unicorn.ts (hunk 16 above)
Comment (davelosert) [reply lib/rules/unicorn.ts]: No reason, i removed it.

File: node.js (hunk 17 above)
Comment (davelosert) [reply node.js]: yep, good catch. Renamed it.

**Comment by davelosert** [comment]
@goloroden : Thanks a lot for the feedback. i put in all your requested changes.

One thing though: You did not leave any feedback for the integration tests.
Did you have a look at those as well?

One Change I did here (after feedback I got from @yeldiRium in Slack): Put all the Test-utilities into the `test/shared` Folder.

File: lib/rules/sharedCoreTypeScript.ts (hunk 13 above)
Comment (goloroden) [reply lib/rules/sharedCoreTypeScript.ts]: What about „The rule camelCase …“?

**Comment by goloroden** [comment]
Meanwhile, I have, and I‘m fine with them. 

So from my POV there is only the workflows and the documentation left.

**Comment by davelosert** [comment]
@goloroden , @yeldiRium - one little update:

 I just added a `todo` comment for the `ecmaVersion` fix that expires on the next `@types/eslint` Version (see [here](https://github.com/thenativeweb/eslint-config-es/pull/433/files#diff-4ffe59dd7ff4e3231ca8cb82496318ec6f9dfbe4d57bfde2c94c4d542d1a22daR20)) as I created a [PR to fix this issue in the `DefinitlyTypes`-Repository](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57037)

**Review by yeldiRium — DISMISSED** [review]

File: lib/rules/sharedCoreTypescript.ts [hunk 21]
```diff
@@ -0,0 +1,200 @@
+import { BetterRulesRecord } from '../betterRules';
```
Comment (goloroden) [inline lib/rules/sharedCoreTypescript.ts:1]: This file is named wrong IMHO – it should be `createSharedRulesFor.ts`, since that's what it exports.

File: test/integration/rules/sharedCoreTypeScriptTests.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
+import { assertLint } from '../../shared/assertLint';
+import { lintJavaScript, lintTsx, lintTypeScript } from '../../shared/esLintTester';
```
Comment (goloroden) [inline test/integration/rules/sharedCoreTypeScriptTests.ts:2]: The functions `lintJavaScript` and `lintTypeScript` need to be renamed to use a lowercase `s`, to conform to the spelling everywhere else.

File: test/shared/esLintTester.ts [hunk 23]
```diff
@@ -0,0 +1,35 @@
+import { ESLint } from 'eslint';
+import path from 'path';
+import * as defaultConfiguration from '../../lib/defaultConfiguration';
+
+const esLintTester = new ESLint({
+  baseConfig: defaultConfiguration,
+  useEslintrc: false
+});
+
+const lintJavaScript = async (codeString: string): Promise<ESLint.LintResult[]> =>
```
Comment (goloroden) [inline test/shared/esLintTester.ts:10]:
```suggestion
const lintJavascript = async (codeString: string): Promise<ESLint.LintResult[]> =>
```

File: test/shared/esLintTester.ts [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+  esLintTester.lintText(codeString, {
+    filePath: 'testFile.js'
+  });
+
+const lintJsx = async (codeString: string): Promise<ESLint.LintResult[]> =>
+  esLintTester.lintText(codeString, {
+    filePath: 'testFile.jsx'
+  });
+
+const lintTypeScript = async (codeString: string): Promise<ESLint.LintResult[]> =>
```
Comment (goloroden) [inline test/shared/esLintTester.ts:20]:
```suggestion
const lintTypescript = async (codeString: string): Promise<ESLint.LintResult[]> =>
```

File: test/unit/betterRules/compileTests.ts [hunk 25]
```diff
@@ -0,0 +1,76 @@
+import { assert } from 'assertthat';
```
Comment (goloroden) [inline test/unit/betterRules/compileTests.ts:1]: Shouldn't this file be named `rulesCompilerTests.ts`?

File: tsconfig.json [hunk 26]
```diff
@@ -0,0 +1,22 @@
...
+    "module": "commonjs",
+    "outDir": "build",
+    "resolveJsonModule": true,
+    "strict": true,
+    "target": "es2019"
+  },
+  "include": [
+    "./lib/**/*.ts",
+    "./test/**/*.ts",
+    "test/shared/dummyTs/dummyTsx.tsx"
```
Comment (goloroden) [inline tsconfig.json:17]: To be consistent with the other paths:

```suggestion
    "./test/shared/dummyTs/dummyTsx.tsx"
```

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @davelosert 👋
Hey @yeldiRium 👋

I've reviewed it once again, and I have a few things left … could you please have a look at the comments?

File: lib/rules/sharedCoreTypescript.ts (hunk 21 above)
Comment (davelosert) [reply lib/rules/sharedCoreTypescript.ts:1]: Note sure if that helps the reader in this case. Especially as there might be other shared rules with other plugins in the future.

Or would you put them then all into this file?

File: test/unit/betterRules/compileTests.ts (hunk 25 above)
Comment (davelosert) [reply test/unit/betterRules/compileTests.ts:1]: The file it tests is named `compile.ts` - shouldn't it be named the same?

**Comment by davelosert** [comment]
👋🏻 Hi @goloroden ,

I had a looked. I fixed some things but have a few questions left on three comments. 🙂

File: test/unit/betterRules/compileTests.ts (hunk 25 above)
Comment (yeldiRium) [reply test/unit/betterRules/compileTests.ts:1]: The top level suit should be `suite('compile')...` ;) This should clear up the confusion.

Comment (goloroden) [reply test/unit/betterRules/compileTests.ts:1]: That's true 👍

File: lib/rules/sharedCoreTypescript.ts (hunk 21 above)
Comment (goloroden) [reply lib/rules/sharedCoreTypescript.ts:1]: Ah okay, now I understand – okay, then this is fine for me 👍

File: test/unit/betterRules/compileTests.ts (hunk 25 above)
Comment (davelosert) [reply test/unit/betterRules/compileTests.ts:1]: Ah yeah, sorry for the confusion. I already fixed it so should be correct now. 👍🏻

**Comment by davelosert** [comment]
Again hi @goloroden & @yeldiRium : I guess everything is fine now and we are ready for merge?

**Review by goloroden — APPROVED** [review]
From my side we are good to go 😊

(I still haven’t reviewed the workflows and the documentation, but this can still be done after merging IMHO)

**Comment by goloroden** [comment]
PS: @yeldiRium If you are fine with this, too, please feel free to merge 😊
